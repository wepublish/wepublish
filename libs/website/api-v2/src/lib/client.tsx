import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
  ApolloLink,
  NormalizedCacheObject
} from '@apollo/client'
import {ComponentType, memo, useMemo} from 'react'
import {BatchHttpLink} from '@apollo/client/link/batch-http'

export const getV2ApiClient = (
  apiUrl: string,
  links: ApolloLink[],
  cache?: NormalizedCacheObject
) => {
  const httpLink = new BatchHttpLink({
    uri: `${apiUrl}/v2`,
    batchMax: 5,
    batchInterval: 20
  })
  const link = [...links, httpLink].reduce(
    (links: ApolloLink | undefined, link) => (links ? links.concat(link) : link),
    undefined
  )

  return new ApolloClient({
    link,
    cache: new InMemoryCache().restore(cache ?? {}),
    ssrMode: typeof window === 'undefined'
  })
}

export const useV2ApiClient = (
  apiUrl: string,
  links: ApolloLink[] = [],
  cache?: NormalizedCacheObject
) => useMemo(() => getV2ApiClient(apiUrl, links, cache), [apiUrl, links, cache])

export const createWithV2ApiClient =
  (apiUrl: string, links: ApolloLink[] = [], cache?: NormalizedCacheObject) =>
  <
    // eslint-disable-next-line @typescript-eslint/ban-types
    P extends object
  >(
    ControlledComponent: ComponentType<P>
  ) =>
    memo<P>(props => {
      const client = useV2ApiClient(apiUrl, links, cache)

      return (
        <ApolloProvider client={client}>
          <ControlledComponent {...props} />
        </ApolloProvider>
      )
    })
