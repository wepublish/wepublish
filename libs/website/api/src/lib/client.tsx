import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
  ApolloLink,
  NormalizedCacheObject
} from '@apollo/client'
import {BatchHttpLink} from '@apollo/client/link/batch-http'

import {ComponentType, memo, useMemo} from 'react'

export const getV1ApiClient = (
  apiUrl: string,
  links: ApolloLink[],
  cache?: NormalizedCacheObject
) => {
  const httpLink = new BatchHttpLink({uri: `${apiUrl}/v1`, batchMax: 5, batchInterval: 20})
  const link = [...links, httpLink].reduce(
    (links: ApolloLink | undefined, link) => (links ? links.concat(link) : link),
    undefined
  )

  return new ApolloClient({
    link,
    cache: new InMemoryCache().restore(cache ?? {}),
    ssrMode: typeof window === 'undefined'
  })
}

export const useV1ApiClient = (
  apiUrl: string,
  links: ApolloLink[] = [],
  cache?: NormalizedCacheObject
) => useMemo(() => getV1ApiClient(apiUrl, links, cache), [apiUrl, links, cache])

export const createWithV1ApiClient =
  (apiUrl: string, links: ApolloLink[] = [], cache?: NormalizedCacheObject) =>
  <
    // eslint-disable-next-line @typescript-eslint/ban-types
    P extends object
  >(
    ControlledComponent: ComponentType<P>
  ) =>
    memo<P>(props => {
      const client = useV1ApiClient(apiUrl, links, cache)

      return (
        <ApolloProvider client={client}>
          <ControlledComponent {...props} />
        </ApolloProvider>
      )
    })
