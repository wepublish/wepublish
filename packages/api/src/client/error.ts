export enum ErrorCode {
  TokenExpired = 'tokenExpired',
  InvalidToken = 'invalidToken',
  Unauthorized = 'unauthorized',
  InvalidCredentials = 'invalidCredentials'
}
