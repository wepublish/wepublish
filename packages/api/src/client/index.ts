export * from './types'
export * from './utility'
export * from './queryArticles'
