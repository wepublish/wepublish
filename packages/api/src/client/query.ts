import axios from 'axios'

const request = axios.create()

export class CancelError extends Error {}

export class CancelToken {
  public source = axios.CancelToken.source()

  cancel() {
    this.source.cancel()
  }
}

export interface RequestOptions {
  cancelToken?: CancelToken
}

export interface QueryOptions {
  query?: string
  variables?: {[key: string]: any}
}

export async function query(
  url: string,
  {query = '', variables}: QueryOptions,
  {cancelToken}: RequestOptions = {}
): Promise<any> {
  try {
    const result = await request.post(
      url,
      {
        query,
        variables
      },
      {
        cancelToken: cancelToken && cancelToken.source.token
      }
    )

    return result.data.data
  } catch (err) {
    if (axios.isCancel(err)) throw new CancelError()
    throw err
  }
}
