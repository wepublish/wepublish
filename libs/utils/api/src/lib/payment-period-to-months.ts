import {PaymentPeriodicity} from '@prisma/client'

export function mapPaymentPeriodToMonths(periodicity: PaymentPeriodicity) {
  switch (periodicity) {
    case PaymentPeriodicity.yearly:
      return 12
    case PaymentPeriodicity.biannual:
      return 6
    case PaymentPeriodicity.quarterly:
      return 3
    case PaymentPeriodicity.monthly:
      return 1
    default:
      throw new Error(`Enum for PaymentPeriodicity ${periodicity} not defined!`)
  }
}
