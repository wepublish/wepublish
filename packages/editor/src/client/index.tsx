import 'core-js/stable'
import 'regenerator-runtime/runtime'

import React from 'react'
import ReactDOM from 'react-dom'
import {ApolloProvider, ApolloClient, ApolloLink, InMemoryCache} from '@apollo/client'
import {createUploadLink} from 'apollo-upload-client'
import {onError} from '@apollo/client/link/error'

import './i18n'

import {ElementID} from '../shared/elementID'
import {ClientSettings} from '../shared/types'
import {RouteProvider} from './route'
import {AuthProvider} from './authContext'
import {LocalStorageKey} from './utility'
import {TwitterProvider} from './blocks/embeds/twitter'
import {InstagramProvider} from './blocks/embeds/instagram'
import {FacebookProvider} from './blocks/embeds/facebook'
import {HotApp} from './app'

// See: https://www.apollographql.com/docs/react/data/fragments/#fragments-on-unions-and-interfaces
export async function fetchIntrospectionQueryResultData(url: string) {
  const response = await fetch(url, {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify({
      variables: {},
      query: `
      {
        __schema {
          types {
            kind
            name
            possibleTypes {
              name
            }
          }
        }
      }
    `
    })
  })

  const result = await response.json()

  const possibleTypes: any = {}

  result.data.__schema.types.forEach((supertype: any) => {
    if (supertype.possibleTypes) {
      possibleTypes[supertype.name] = supertype.possibleTypes.map((subtype: any) => subtype.name)
    }
  })

  return possibleTypes
}

const onDOMContentLoaded = async () => {
  const {apiURL}: ClientSettings = JSON.parse(
    document.getElementById(ElementID.Settings)!.textContent!
  )

  const adminAPIURL = `${apiURL}/admin`

  const authLink = new ApolloLink((operation, forward) => {
    const token = localStorage.getItem(LocalStorageKey.SessionToken)
    const context = operation.getContext()

    operation.setContext({
      headers: {
        authorization: token ? `Bearer ${token}` : '',
        ...context.headers
      },
      ...context
    })

    return forward(operation)
  })

  const authErrorLink = onError(({graphQLErrors, /* networkError, */ operation, forward}) => {
    if (graphQLErrors) {
      graphQLErrors.forEach(({/* message, locations, path, */ extensions}) => {
        // console.log(`[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`)
        if (extensions?.code === 'UNAUTHENTICATED') {
          window.location.pathname = '/logout'
        }
      })
      // if (networkError) console.log(`[Network error]: ${networkError}`)
    }
    return forward(operation)
  })

  const client = new ApolloClient({
    link: authLink.concat(authErrorLink).concat(createUploadLink({uri: adminAPIURL})),
    cache: new InMemoryCache({
      possibleTypes: await fetchIntrospectionQueryResultData(adminAPIURL)
    })
  })

  window.addEventListener('dragover', e => e.preventDefault())
  window.addEventListener('drop', e => e.preventDefault())

  ReactDOM.render(
    <ApolloProvider client={client}>
      <AuthProvider>
        <RouteProvider>
          <FacebookProvider sdkLanguage={'en_US'}>
            <InstagramProvider>
              <TwitterProvider>
                <HotApp />
              </TwitterProvider>
            </InstagramProvider>
          </FacebookProvider>
        </RouteProvider>
      </AuthProvider>
    </ApolloProvider>,
    document.getElementById(ElementID.ReactRoot)
  )
}

if (document.readyState !== 'loading') {
  onDOMContentLoaded()
} else {
  document.addEventListener('DOMContentLoaded', onDOMContentLoaded)
}
