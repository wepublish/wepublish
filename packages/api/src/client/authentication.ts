import {query, RequestOptions} from './query'

export async function authenticateWithCredentials(
  url: string,
  email: string,
  password: string,
  opts?: RequestOptions
): Promise<{
  user: {email: string}
  refreshToken: string
  accessToken: string
  refreshTokenExpiresIn: number
  accessTokenExpiresIn: number
}> {
  const {authenticateWithCredentials: response} = await query(
    url,
    {
      query: `
    mutation authenticateWithCredentials($email: String!, $password: String!) {
      authenticateWithCredentials(email: $email, password: $password) {
        user {email},
        refreshToken,
        accessToken,
        refreshTokenExpiresIn,
        accessTokenExpiresIn
      }
    }
  `,
      variables: {
        email,
        password
      }
    },
    opts
  )
  return response
}
