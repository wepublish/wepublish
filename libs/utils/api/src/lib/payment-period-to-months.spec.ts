import {PaymentPeriodicity} from '@prisma/client'
import {mapPaymentPeriodToMonths} from './payment-period-to-months'

describe('mapPaymentPeriodToMonths', () => {
  it.each([
    PaymentPeriodicity.monthly,
    PaymentPeriodicity.quarterly,
    PaymentPeriodicity.biannual,
    PaymentPeriodicity.yearly
  ])('should format payment period for %s', periodicity => {
    expect(mapPaymentPeriodToMonths(periodicity)).toMatchSnapshot()
  })

  it('should error if if periodicity is not found', () => {
    expect(() => mapPaymentPeriodToMonths('invalid' as any)).toThrowError()
  })
})
