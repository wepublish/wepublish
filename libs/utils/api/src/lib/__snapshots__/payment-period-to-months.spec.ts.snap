// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapPaymentPeriodToMonths should format payment period for biannual 1`] = `6`;

exports[`mapPaymentPeriodToMonths should format payment period for monthly 1`] = `1`;

exports[`mapPaymentPeriodToMonths should format payment period for quarterly 1`] = `3`;

exports[`mapPaymentPeriodToMonths should format payment period for yearly 1`] = `12`;
